lib/streamlit/commands/navigation.py

```diff
@@ -0,0 +1,178 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Literal
+
+from typing_extensions import TypeAlias
+
+from streamlit.errors import StreamlitAPIException
+from streamlit.navigation.page import StreamlitPage
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+from streamlit.runtime.metrics_util import gather_metrics
+from streamlit.runtime.scriptrunner.script_run_context import (
+    ScriptRunContext,
+    get_script_run_ctx,
+)
+from streamlit.source_util import PageHash, PageInfo
+
+SectionHeader: TypeAlias = str
+
+
+def pages_from_nav_sections(
+    nav_sections: dict[SectionHeader, list[StreamlitPage]]
+) -> list[StreamlitPage]:
+    page_list = []
+    for pgs in nav_sections.values():
+        for page in pgs:
+            page_list.append(page)
+
+    return page_list
+
+
+def nav_sections_to_page_dict(
+    nav_sections: dict[SectionHeader, list[StreamlitPage]]
+) -> dict[PageHash, PageInfo]:
+    d: dict[PageHash, PageInfo] = {}
+    for section_header in nav_sections:
+        for page in nav_sections[section_header]:
+            if isinstance(page._page, Path):
+                script_path = str(page._page)
+            else:
+                script_path = ""
+            d[page._script_hash] = {
+                "page_script_hash": page._script_hash,
+                "page_name": page.title,
+                "icon": page.icon,
+                "script_path": script_path,
+                "url_pathname": page.title.replace(" ", "_"),
+            }
+    return d
+
+
+def send_page_not_found(ctx: ScriptRunContext):
+    msg = ForwardMsg()
+    msg.page_not_found.page_name = ""
+    ctx.enqueue(msg)
+
+
+@gather_metrics("navigation")
+def navigation(
+    pages: list[StreamlitPage] | dict[str, list[StreamlitPage]],
```

Comment:
nit: should the key-type be `SectionHeader` instead of `str`?

Reply:
I would be fine, but do we need to worry for the docs that come from the types. I'll follow up.